File: iosApp/Ehon/ScenePainter.swift
import UIKit
import EhonCore

/// Walks a `Scene` onto a CoreGraphics context.
///
/// Deliberately the structural twin of `painter-compose/ScenePainter.kt`: no layout, no
/// measurement, no decisions — every coordinate was resolved by the shared `SceneBuilder`.
/// That is what makes the screen, the 2048px share image and the 300dpi PDF the same
/// picture, and what keeps this file and its Compose counterpart from being able to drift
/// on anything except rasterisation.
///
/// Assumes a y-down context with a top-left origin, which is what UIKit's renderers and
/// SwiftUI's `withCGContext` both provide.
struct ScenePainter {

    let imageProvider: (String) -> UIImage?

    init(imageProvider: @escaping (String) -> UIImage? = { _ in nil }) {
        self.imageProvider = imageProvider
    }

    func draw(_ scene: Scene, into ctx: CGContext) {
        let bounds = CGRect(x: 0, y: 0, width: CGFloat(scene.size.w), height: CGFloat(scene.size.h))

        ctx.saveGState()
        if scene.cornerRadius > 0 {
            let rounded = UIBezierPath(roundedRect: bounds, cornerRadius: CGFloat(scene.cornerRadius))
            ctx.addPath(rounded.cgPath)
            ctx.clip()
        }
        ctx.setFillColor(scene.background.cgColor)
        ctx.fill(bounds)

        for node in scene.nodes {
            draw(node, into: ctx, pageBounds: bounds)
        }
        ctx.restoreGState()
    }

    private func draw(_ node: any SceneNode, into ctx: CGContext, pageBounds: CGRect) {
        switch node {

        case let group as SceneNodeGroup:
            ctx.saveGState()
            ctx.translateBy(x: CGFloat(group.pivot.x), y: CGFloat(group.pivot.y))
            ctx.rotate(by: CGFloat(group.rotationDeg) * .pi / 180)
            ctx.translateBy(x: -CGFloat(group.pivot.x), y: -CGFloat(group.pivot.y))
            for child in group.children {
                draw(child, into: ctx, pageBounds: pageBounds)
            }
            ctx.restoreGState()

        case let ellipse as SceneNodeEllipse:
            ctx.setFillColor(ellipse.fill.cgColor)
            ctx.fillEllipse(in: ellipse.rect.cgRect)
            if ellipse.hasHairline {
                ctx.setStrokeColor(Self.hairlineColour)
                ctx.setLineWidth(Self.hairlineWidth)
                ctx.strokeEllipse(in: ellipse.rect.cgRect)
            }

        case let rounded as SceneNodeRoundRect:
            let path = Self.roundRectPath(rounded.rect.cgRect, radii: rounded.radii.map { CGFloat(truncating: $0) })
            ctx.setFillColor(rounded.fill.cgColor)
            ctx.addPath(path)
            ctx.fillPath()
            if rounded.hasHairline {
                ctx.setStrokeColor(Self.hairlineColour)
                ctx.setLineWidth(Self.hairlineWidth)
                ctx.addPath(path)
                ctx.strokePath()
            }

        case let polygon as SceneNodePolygon:
            guard let first = polygon.points.first else { break }
            let path = CGMutablePath()
            path.move(to: first.cgPoint)
            for point in polygon.points.dropFirst() { path.addLine(to: point.cgPoint) }
            path.closeSubpath()
            ctx.setFillColor(polygon.fill.cgColor)
            ctx.addPath(path)
            ctx.fillPath()
            if polygon.hasHairline {
                ctx.setStrokeColor(Self.hairlineColour)
                ctx.setLineWidth(Self.hairlineWidth)
                ctx.addPath(path)
                ctx.strokePath()
            }

        // The three shapes that were CSS radial-gradient masks in the prototype.
        // Even-odd fills instead: no mask layer, and crisp at 300dpi.
        case let ring as SceneNodeRing:
            ctx.setFillColor(ring.fill.cgColor)
            ctx.addPath(Self.ringPath(ring.rect.cgRect, innerRatio: CGFloat(ring.innerRatio)))
            ctx.fillPath(using: .evenOdd)

        case let arc as SceneNodeArc:
            let full = arc.rect.cgRect
            ctx.saveGState()
            ctx.clip(to: CGRect(x: full.minX, y: full.minY, width: full.width, height: full.height / 2))
            ctx.setFillColor(arc.fill.cgColor)
            ctx.addPath(Self.ringPath(full, innerRatio: CGFloat(arc.innerRatio)))
            ctx.fillPath(using: .evenOdd)
            ctx.restoreGState()

        case let crescent as SceneNodeCrescent:
            // CoreGraphics has no path subtraction, so clip to the body and then clip
            // away the cutting circle with an even-odd rect-minus-circle.
            let body = crescent.rect.cgRect
            let radius = CGFloat(crescent.cutRadius)
            let cut = CGRect(
                x: CGFloat(crescent.cutCentre.x) - radius,
                y: CGFloat(crescent.cutCentre.y) - radius,
                width: radius * 2,
                height: radius * 2
            )
            ctx.saveGState()
            ctx.addEllipse(in: body)
            ctx.clip()
            let outer = body.union(cut).insetBy(dx: -1, dy: -1)
            let mask = CGMutablePath()
            mask.addRect(outer)
            mask.addEllipse(in: cut)
            ctx.addPath(mask)
            ctx.clip(using: .evenOdd)
            ctx.setFillColor(crescent.fill.cgColor)
            ctx.fill(body)
            ctx.restoreGState()

        case let image as SceneNodeImage:
            guard let bitmap = imageProvider(image.assetName), let cgImage = bitmap.cgImage else { break }
            ctx.saveGState()
            // UIImage draws y-up in a CGContext, so flip within the destination rect.
            let rect = image.rect.cgRect
            ctx.translateBy(x: rect.minX, y: rect.minY + rect.height)
            ctx.scaleBy(x: 1, y: -1)
            ctx.draw(cgImage, in: CGRect(x: 0, y: 0, width: rect.width, height: rect.height))
            ctx.restoreGState()

        case let text as SceneNodeText:
            if let ruby = text.ruby, let origin = text.rubyOrigin {
                Self.drawString(ruby, at: origin.cgPoint, size: CGFloat(text.rubySizePx),
                                fill: text.fill, role: text.font)
            }
            Self.drawString(text.base, at: text.baseOrigin.cgPoint, size: CGFloat(text.baseSizePx),
                            fill: text.fill, role: text.font)

        case let stroke as SceneNodeStrokePath:
            guard let first = stroke.points.first else { break }
            let path = CGMutablePath()
            path.move(to: first.cgPoint)
            if stroke.points.count == 1 {
                // A tap is a dot; a hair of length makes the round cap render.
                path.addLine(to: CGPoint(x: first.cgPoint.x + 0.01, y: first.cgPoint.y))
            } else {
                for point in stroke.points.dropFirst() { path.addLine(to: point.cgPoint) }
            }
            ctx.saveGState()
            ctx.setLineCap(.round)
            ctx.setLineJoin(.round)
            ctx.setLineWidth(CGFloat(stroke.widthPx))
            // The eraser lifts pixels rather than painting over them, so erasing above a
            // part reveals the page instead of smearing the page colour on top.
            ctx.setBlendMode(stroke.erase ? .clear : .normal)
            ctx.setStrokeColor(stroke.fill.cgColor)
            ctx.addPath(path)
            ctx.strokePath()
            ctx.restoreGState()

        case let selection as SceneNodeSelectionRing:
            ctx.saveGState()
            ctx.setStrokeColor(selection.stroke.cgColor)
            ctx.setLineWidth(Self.selectionWidth)
            ctx.setLineDash(phase: 0, lengths: [4, 3])
            ctx.addPath(UIBezierPath(roundedRect: selection.rect.cgRect,
                                     cornerRadius: Self.selectionRadius).cgPath)
            ctx.strokePath()
            ctx.restoreGState()

        default:
            assertionFailure("unhandled SceneNode: \(type(of: node))")
        }
    }

    // MARK: - helpers

    private static func drawString(
        _ text: String, at origin: CGPoint, size: CGFloat, fill: Int32, role: FontRole
    ) {
        // NSString.draw(at:) takes a TOP-LEFT origin and honours the y-down context, which
        // is exactly the convention SceneBuilder emits. Same call the measurer uses.
        (text as NSString).draw(at: origin, withAttributes: [
            .font: EhonFonts.font(for: role, size: size),
            .foregroundColor: UIColor(cgColor: fill.cgColor),
        ])
    }

    private static func ringPath(_ rect: CGRect, innerRatio: CGFloat) -> CGPath {
        let path = CGMutablePath()
        path.addEllipse(in: rect)
        let inner = CGRect(
            x: rect.midX - rect.width * innerRatio / 2,
            y: rect.midY - rect.height * innerRatio / 2,
            width: rect.width * innerRatio,
            height: rect.height * innerRatio
        )
        path.addEllipse(in: inner)
        return path
    }

    private static func roundRectPath(_ rect: CGRect, radii: [CGFloat]) -> CGPath {
        let cap = min(rect.width, rect.height) / 2
        let r = radii.map { min($0, cap) }
        guard r.count == 4 else {
            return UIBezierPath(roundedRect: rect, cornerRadius: r.first ?? 0).cgPath
        }
        let path = CGMutablePath()
        // Clockwise from top-left, matching SceneBuilder's radii order.
        path.move(to: CGPoint(x: rect.minX + r[0], y: rect.minY))
        path.addLine(to: CGPoint(x: rect.maxX - r[1], y: rect.minY))
        path.addArc(tangent1End: CGPoint(x: rect.maxX, y: rect.minY),
                    tangent2End: CGPoint(x: rect.maxX, y: rect.minY + r[1]), radius: r[1])
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY - r[2]))
        path.addArc(tangent1End: CGPoint(x: rect.maxX, y: rect.maxY),
                    tangent2End: CGPoint(x: rect.maxX - r[2], y: rect.maxY), radius: r[2])
        path.addLine(to: CGPoint(x: rect.minX + r[3], y: rect.maxY))
        path.addArc(tangent1End: CGPoint(x: rect.minX, y: rect.maxY),
                    tangent2End: CGPoint(x: rect.minX, y: rect.maxY - r[3]), radius: r[3])
        path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + r[0]))
        path.addArc(tangent1End: CGPoint(x: rect.minX, y: rect.minY),
                    tangent2End: CGPoint(x: rect.minX + r[0], y: rect.minY), radius: r[0])
        path.closeSubpath()
        return path
    }

    /// A single design constant, so scene nodes carry only a flag.
    private static let hairlineColour: CGColor = Organic.shared.hairline.cgColor

    private static let hairlineWidth: CGFloat = 1
    private static let selectionRadius: CGFloat = 6
    private static let selectionWidth: CGFloat = 2
}

// MARK: - bridging

/// `Argb` is a Kotlin inline value class, so it crosses the Obj-C boundary as a packed
/// `Int32` rather than a type. That avoids boxing a colour per shape, at the cost of
/// needing this one conversion.
extension Int32 {
    var cgColor: CGColor {
        let packed = UInt32(bitPattern: self)
        return CGColor(
            srgbRed: CGFloat((packed >> 16) & 0xFF) / 255,
            green: CGFloat((packed >> 8) & 0xFF) / 255,
            blue: CGFloat(packed & 0xFF) / 255,
            alpha: CGFloat((packed >> 24) & 0xFF) / 255
        )
    }
}

extension EhonCore.Rect {
    var cgRect: CGRect {
        CGRect(x: CGFloat(x), y: CGFloat(y), width: CGFloat(w), height: CGFloat(h))
    }
}

extension EhonCore.Point {
    var cgPoint: CGPoint { CGPoint(x: CGFloat(x), y: CGFloat(y)) }
}

File: painter-compose/src/commonMain/kotlin/app/ehon/painter/ScenePainter.kt
package app.ehon.painter

import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect as ComposeRect
import androidx.compose.ui.geometry.RoundRect
import androidx.compose.ui.geometry.Size as ComposeSize
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathFillType
import androidx.compose.ui.graphics.PathOperation
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke as DrawStroke
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.graphics.drawscope.clipRect
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.graphics.drawscope.translate
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.TextUnitType
import app.ehon.design.Argb
import app.ehon.design.Organic
import app.ehon.geom.Rect
import app.ehon.model.FontFace
import app.ehon.scene.Scene
import app.ehon.scene.SceneNode
import kotlin.math.min

/**
 * Walks a [Scene] onto a Compose [DrawScope].
 *
 * Deliberately thin: it contains no layout, no measurement and no decisions. Everything
 * geometric was resolved by `SceneBuilder`, which is why the iOS CoreGraphics painter can
 * be a structurally identical file and why the two cannot drift on anything but
 * rasterisation.
 */
class ScenePainter(
    private val textMeasurer: TextMeasurer,
    private val bodyFont: FontFamily,
    private val uiFont: FontFamily,
    private val imageProvider: ImageProvider,
) {

    fun DrawScope.paint(scene: Scene) {
        clipPage(scene) {
            drawRect(color = scene.background.toColor())
            scene.nodes.forEach { node -> draw(node) }
        }
    }

    private inline fun DrawScope.clipPage(scene: Scene, crossinline body: DrawScope.() -> Unit) {
        if (scene.cornerRadius <= 0f) {
            body()
            return
        }
        val rounded = Path().apply {
            addRoundRect(
                RoundRect(
                    ComposeRect(0f, 0f, scene.size.w, scene.size.h),
                    CornerRadius(scene.cornerRadius),
                ),
            )
        }
        clipPath(rounded) { body() }
    }

    private fun DrawScope.draw(node: SceneNode) {
        when (node) {
            is SceneNode.Group ->
                rotate(node.rotationDeg, Offset(node.pivot.x, node.pivot.y)) {
                    node.children.forEach { draw(it) }
                }

            is SceneNode.Ellipse -> {
                drawOval(node.fill.toColor(), node.rect.offset(), node.rect.size())
                if (node.hasHairline) {
                    drawOval(
                        Organic.hairline.toColor(), node.rect.offset(), node.rect.size(),
                        style = hairline(),
                    )
                }
            }

            is SceneNode.RoundRect -> {
                val path = roundRectPath(node.rect, node.radii)
                drawPath(path, node.fill.toColor())
                if (node.hasHairline) drawPath(path, Organic.hairline.toColor(), style = hairline())
            }

            is SceneNode.Polygon -> {
                val path = Path().apply {
                    node.points.forEachIndexed { i, p ->
                        if (i == 0) moveTo(p.x, p.y) else lineTo(p.x, p.y)
                    }
                    close()
                }
                drawPath(path, node.fill.toColor())
                if (node.hasHairline) drawPath(path, Organic.hairline.toColor(), style = hairline())
            }

            // The three former CSS radial-gradient masks. Even-odd rather than masking:
            // fewer moving parts, and it stays sharp at 300dpi.
            is SceneNode.Ring -> drawPath(ringPath(node.rect, node.innerRatio), node.fill.toColor())

            is SceneNode.Arc -> {
                clipRect(
                    node.rect.x,
                    node.rect.y,
                    node.rect.x + node.rect.w,
                    node.rect.y + node.rect.h / 2f,
                ) {
                    drawPath(ringPath(node.rect, node.innerRatio), node.fill.toColor())
                }
            }

            is SceneNode.Crescent -> {
                val body = Path().apply { addOval(ComposeRect(node.rect.offset(), node.rect.size())) }
                val cut = Path().apply {
                    addOval(
                        ComposeRect(
                            center = Offset(node.cutCentre.x, node.cutCentre.y),
                            radius = node.cutRadius,
                        ),
                    )
                }
                drawPath(Path().apply { op(body, cut, PathOperation.Difference) }, node.fill.toColor())
            }

            is SceneNode.Image ->
                imageProvider.load(node.assetName)?.let { image ->
                    translate(node.rect.x, node.rect.y) {
                        drawImage(
                            image = image,
                            dstSize = androidx.compose.ui.unit.IntSize(
                                node.rect.w.toInt(),
                                node.rect.h.toInt(),
                            ),
                        )
                    }
                }

            is SceneNode.Text -> {
                node.ruby?.let { ruby ->
                    drawString(ruby, node.rubyOrigin!!, node.rubySizePx, node.fill, node.font)
                }
                drawString(node.base, node.baseOrigin, node.baseSizePx, node.fill, node.font)
            }

            is SceneNode.StrokePath -> {
                if (node.points.isEmpty()) return
                val path = Path().apply {
                    val first = node.points.first()
                    moveTo(first.x, first.y)
                    if (node.points.size == 1) {
                        // A tap is a dot; give it a hair of length so it renders round.
                        lineTo(first.x + 0.01f, first.y)
                    } else {
                        node.points.drop(1).forEach { lineTo(it.x, it.y) }
                    }
                }
                drawPath(
                    path = path,
                    color = node.fill.toColor(),
                    style = DrawStroke(node.widthPx, cap = StrokeCap.Round, join = StrokeJoin.Round),
                    // Eraser lifts pixels rather than painting over them, so a stroke
                    // erased above a part reveals the page, not a coloured smear.
                    blendMode = if (node.erase) BlendMode.Clear else BlendMode.SrcOver,
                )
            }

            is SceneNode.SelectionRing ->
                drawRoundRect(
                    color = node.stroke.toColor(),
                    topLeft = node.rect.offset(),
                    size = node.rect.size(),
                    cornerRadius = CornerRadius(SELECTION_RADIUS),
                    style = DrawStroke(SELECTION_WIDTH),
                )
        }
    }

    private fun DrawScope.drawString(
        text: String,
        origin: app.ehon.geom.Point,
        sizePx: Float,
        fill: Argb,
        face: FontFace,
    ) {
        drawText(
            textMeasurer = textMeasurer,
            text = text,
            topLeft = Offset(origin.x, origin.y),
            style = TextStyle(
                color = fill.toColor(),
                fontSize = TextUnit(sizePx, TextUnitType.Sp),
                fontFamily = if (face == FontFace.UI) uiFont else bodyFont,
            ),
        )
    }

    private fun ringPath(rect: Rect, innerRatio: Float) = Path().apply {
        fillType = PathFillType.EvenOdd
        addOval(ComposeRect(rect.offset(), rect.size()))
        val inset = Rect(
            x = rect.centerX - rect.w * innerRatio / 2f,
            y = rect.centerY - rect.h * innerRatio / 2f,
            w = rect.w * innerRatio,
            h = rect.h * innerRatio,
        )
        addOval(ComposeRect(inset.offset(), inset.size()))
    }

    private fun roundRectPath(rect: Rect, radii: List<Float>): Path {
        val cap = min(rect.w, rect.h) / 2f
        val r = radii.map { CornerRadius(it.coerceAtMost(cap)) }
        return Path().apply {
            addRoundRect(
                RoundRect(
                    rect = ComposeRect(rect.offset(), rect.size()),
                    topLeft = r[0],
                    topRight = r[1],
                    bottomRight = r[2],
                    bottomLeft = r[3],
                ),
            )
        }
    }

    private fun hairline() = DrawStroke(HAIRLINE_WIDTH)

    private fun Rect.offset() = Offset(x, y)

    private fun Rect.size() = ComposeSize(w, h)

    private fun Argb.toColor() = Color(r, g, b, a)

    private companion object {
        const val HAIRLINE_WIDTH = 1f
        const val SELECTION_RADIUS = 6f
        const val SELECTION_WIDTH = 2f
    }
}

/** Resolves a raster part's asset name to a decoded, cached bitmap. */
fun interface ImageProvider {
    fun load(assetName: String): androidx.compose.ui.graphics.ImageBitmap?
}
